src/main/java/org/tron/common/runtime/vm/program/VMTransaction.java

```diff
@@ -198,194 +207,214 @@ public synchronized void protoParse() {
+    this.data = data;
+    parsed = true;
+  }
+
+  public ECDSASignature getSignature() {
+    protoParse();
+    return signature;
+  }
+
+  public byte[] getContractAddress() {
+    if (!isContractCreation()) {
+      return null;
+    }
+    // return HashUtil.calcNewAddr(this.getSender(), this.getNonce());
+    DataWord addr = new DataWord(this.getSender());
+    byte[] privKey = Sha256Hash.hash(addr.getData());
+    ECKey ecKey = ECKey.fromPrivate(privKey);
+    byte[] newAddress = ecKey.getAddress();
+    return newAddress;
+  }
+
+  public boolean isContractCreation() {
+    protoParse();
+    return this.receiveAddress == null || Arrays
+        .equals(this.receiveAddress, ByteUtil.EMPTY_BYTE_ARRAY);
+  }
+
+  /*
+   * Crypto
+   */
+
+  public ECKey getKey() {
+    byte[] hash = getRawHash();
+    return ECKey.recoverFromSignature(signature.v, signature, hash);
+  }
+
+  public synchronized byte[] getSender() {
+    try {
+      if (sendAddress == null && getSignature() != null) {
+        sendAddress = ECKey.signatureToAddress(getRawHash(), getSignature());
+      }
+      return sendAddress;
+    } catch (SignatureException e) {
+      logger.error(e.getMessage(), e);
+    }
+    return null;
+  }
+
+  public Integer getChainId() {
+    protoParse();
+    return chainId == null ? null : (int) chainId;
+  }
+
+  /**
+   * @deprecated should prefer #sign(ECKey) over this method
+   */
+  public void sign(byte[] privKeyBytes) throws MissingPrivateKeyException {
+    sign(ECKey.fromPrivate(privKeyBytes));
+  }
+
+  public void sign(ECKey key) throws MissingPrivateKeyException {
+    this.signature = key.sign(this.getRawHash());
+    this.protoEncoded = null;
+  }
+
+  @Override
+  public String toString() {
+    return toString(Integer.MAX_VALUE);
+  }
+
+  public String toString(int maxDataSize) {
+    protoParse();
+    String dataS;
+    if (data == null) {
+      dataS = "";
+    } else if (data.length < maxDataSize) {
+      dataS = ByteUtil.toHexString(data);
+    } else {
+      dataS = ByteUtil.toHexString(Arrays.copyOfRange(data, 0, maxDataSize)) +
+          "... (" + data.length + " bytes)";
+    }
+    return "TransactionData [" + "hash=" + ByteUtil.toHexString(hash) +
+        ", receiveAddress=" + ByteUtil.toHexString(receiveAddress) +
+        ", sendAddress=" + ByteUtil.toHexString(getSender()) +
+        ", value=" + value +
+        ", data=" + dataS +
+        ", signatureV=" + (signature == null ? "" : signature.v) +
+        ", signatureR=" + (signature == null ? ""
+        : ByteUtil.toHexString(BigIntegers.asUnsignedByteArray(signature.r))) +
+        ", signatureS=" + (signature == null ? ""
+        : ByteUtil.toHexString(BigIntegers.asUnsignedByteArray(signature.s))) +
+        "]";
+  }
+
+  /**
+   * For signatures you have to keep also RLP of the transaction without any signature data
+   */
+  public byte[] getEncodedRaw() {
+    protoParse();
+    if (protoRaw != null) {
+      return protoRaw;
```

Comment:
![Vulnerability](https://sonarsource.github.io/sonar-github/vulnerability.png 'Vulnerability') Vulnerability: Return a copy of "protoRaw". ([squid:S2384](https://sonarcloud.io/organizations/tron-zhaohong/rules#rule_key=squid%3AS2384))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=java-tron&pullRequest=1188&issues=AWTf_bGjrFWWQMKkhNw_&open=AWTf_bGjrFWWQMKkhNw_)

[ISSUE_KEY]: AWTf_bGjrFWWQMKkhNw_